sql/core/src/main/scala/org/apache/spark/sql/execution/ExplainUtils.scala

```diff
@@ -0,0 +1,214 @@
+      }
+    } catch {
+      case e: AnalysisException => append(e.toString)
+    }
+    currentOperatorID
+  }
+
+  /**
+   * Given a input physical plan, performs the following tasks.
+   *   1. Generates the explain output for the input plan excluding the subquery plans.
+   *   2. Generates the explain output for each subquery referenced in the plan.
+   */
+  def processPlan[T <: QueryPlan[T]](
+      plan: => QueryPlan[T],
+      append: String => Unit): Unit = {
+    try {
+      val subqueries = ArrayBuffer.empty[(SparkPlan, Expression, SparkPlan)]
+      var currentOperatorID = 0
+      currentOperatorID = processPlanSkippingSubqueries(plan, append, currentOperatorID)
+      getSubqueries(plan, subqueries)
+      var i = 0
+
+      // Process all the subqueries in the plan.
+      for (sub <- subqueries) {
+        if (i == 0) {
+          append("\n===== Subqueries =====\n\n")
+        }
+        i = i + 1
+        append(s"Subquery:$i Hosting operator id = " +
+          s"${getOpId(sub._1)} Hosting Expression = ${sub._2}\n")
+
+        // For each subquery expression in the parent plan, process its child plan to compute
+        // the explain output.
+        currentOperatorID = processPlanSkippingSubqueries(
+          sub._3,
+          append,
+          currentOperatorID)
+        append("\n")
+      }
+    } finally {
+      removeTags(plan)
+    }
+  }
+
+  /**
+   * Traverses the supplied input plan in a bottem-up fashion to produce the following two maps :
+   *    1. operator -> operator identifier
+   *    2. operator identifier -> operator
+   * Note :
+   *    1. Operator such as WholeStageCodegenExec and InputAdapter are skipped as they don't
+   *       appear in the explain output.
+   *    2. operator identifier starts at startIdx + 1
+   */
+  private def generateOperatorIDs(
+      plan: QueryPlan[_],
+      startOperatorID: Int,
+      operatorIDs: mutable.ArrayBuffer[(Int, QueryPlan[_])]): Int = {
+    var currentOperationID = startOperatorID
+    // Skip the subqueries as they are not printed as part of main query block.
+    if (plan.isInstanceOf[BaseSubqueryExec]) {
+      return currentOperationID
+    }
+    plan.foreachUp {
+      case p: WholeStageCodegenExec =>
+      case p: InputAdapter =>
+      case other: QueryPlan[_] =>
+        if (!other.getTagValue(QueryPlan.opidTag).isDefined) {
+          currentOperationID += 1
+          other.setTagValue(QueryPlan.opidTag, currentOperationID)
+          operatorIDs += ((currentOperationID, other))
+        }
+        other.innerChildren.foreach { plan =>
+          currentOperationID = generateOperatorIDs(plan,
+            currentOperationID,
+            operatorIDs)
+        }
+    }
+    currentOperationID
+  }
+  /**
+   * Traverses the supplied input plan in a top-down fashion to produce the following map:
+   *    1. operator -> whole stage codegen id
+   */
+  private def generateWholeStageCodegenIdMap(plan: QueryPlan[_]): Unit = {
```

Comment:
This method doesn't generate a map, we should update the method name.

Reply:
@cloud-fan thanks.. will change.